### PR TITLE
refactor set rate

### PR DESCRIPTION
**Pull request by cyberliem**

### CODE REVIEW

**Comment by cyberliem** [comment]
[Refactor logic flow in SetRates function](https://trello.com/c/agGT4IdL/172-refactor-logic-flow-in-setrates-function)

File: core/reserve_core.go [hunk 1]
```diff
@@ -339,6 +339,65 @@ func (self ReserveCore) pendingSetrateInfo(minedNonce uint64) (*big.Int, *big.In
 	return big.NewInt(int64(nonce)), big.NewInt(int64(gasPrice)), count, nil
 }
 
+func (self ReserveCore) GetSetRateResult(tokens []common.Token,
+	buys []*big.Int,
+	sells []*big.Int,
+	block *big.Int,
+	afpMids []*big.Int) (error, *types.Transaction) {
```
Comment (favadi) [inline core/reserve_core.go]: the function signature could be changed a bit so we can write:

```go
func (self ReserveCore) GetSetRateResult(tokens []common.Token,
	buys, sells, afpMids []*big.Int,
	block *big.Int) (error, *types.Transaction) {
```

Comment (favadi) [reply core/reserve_core.go]: Actually, the error should be second value in return, please change it to.

```go
func (self ReserveCore) GetSetRateResult(tokens []common.Token,
	buys, sells, afpMids []*big.Int,
	block *big.Int) (*types.Transaction, error) {
```

File: core/reserve_core.go [hunk 2]
```diff
@@ -339,6 +339,65 @@ func (self ReserveCore) pendingSetrateInfo(minedNonce uint64) (*big.Int, *big.In
...
+	var (
+		tx  *types.Transaction
+		err error
+	)
+	if len(tokens) != len(buys) || len(tokens) != len(sells) || len(tokens) != len(afpMids) {
+		return errors.New("Tokens, buys sells and afpMids must have the same length"), tx
+	}
```
Comment (favadi) [inline core/reserve_core.go]: I know it is repetitive but please change this to something like:

```go
if len(tokens) != len(buys) {
return fmt.Errorf("number of buys (%d) is not equal to number of tokens (%d), len(buys), len(tokens)
```

This way user will have a better error message.

File: core/reserve_core.go [hunk 3]
```diff
@@ -339,6 +339,65 @@ func (self ReserveCore) pendingSetrateInfo(minedNonce uint64) (*big.Int, *big.In
...
+
+	if err = sanityCheck(buys, afpMids, sells); err != nil {
+		return err, tx
+	}
+	tokenAddrs := []ethereum.Address{}
```
Comment (favadi) [inline core/reserve_core.go]: We should use `var tokenAddrs []ethereum.Address` instead. Reference: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices.

File: core/reserve_core.go [hunk 4]
```diff
@@ -339,6 +339,65 @@ func (self ReserveCore) pendingSetrateInfo(minedNonce uint64) (*big.Int, *big.In
...
+	for _, token := range tokens {
+		tokenAddrs = append(tokenAddrs, ethereum.HexToAddress(token.Address))
+	}
+	// if there is a pending set rate tx, we replace it
+	var oldNonce *big.Int
+	var oldPrice *big.Int
+	var minedNonce uint64
+	var count uint64
```
Comment (favadi) [inline core/reserve_core.go]: those should be grouped  in form

```go
var (
xxx
yyy
)
```

**Pull request merged**